scintillometry/base/tests/test_base.py

```diff
@@ -0,0 +1,136 @@
+
+        super().__init__(ih, shape, sample_rate, samples_per_frame, ih.dtype)
+
+    def _read_frame(self, frame_index):
+        self.ih.seek(frame_index * self._raw_frame_len)
+        return self.ih.read(self._raw_frame_len).reshape((-1,) +
+                                                         self.sample_shape)
+
+
+class TestTaskBase(object):
+
+    @staticmethod
+    def convert_time_offset(offset, sample_rate):
+        return int((offset * sample_rate).to(u.one).round())
+
+    @pytest.mark.parametrize(('n', 'samples_per_frame'),
+                             tuple(itertools.product([256, 337], [1, 7, 16])))
+    def test_taskbase(self, n, samples_per_frame):
+        """Test properties and methods of TaskBase, including
+        self-consistency with varying ``n`` and ``samples_per_frame``.
+        """
+        fh = vdif.open(SAMPLE_VDIF)
+        rt = ReshapeTask(fh, n, samples_per_frame=samples_per_frame)
+
+        # Check sample pointer.
+        assert rt.sample_rate == fh.sample_rate / n
+        nsample = samples_per_frame * (fh.shape[0] // n // samples_per_frame)
+        assert rt.shape == (nsample, n) + fh.sample_shape
+        assert rt.size == np.prod(rt.shape)
+        assert rt.ndim == fh.ndim + 1
+        assert rt.tell() == 0
+        assert rt.tell(unit='time') == rt.time == rt.start_time
+        assert abs(rt.stop_time - rt.start_time -
+                   (nsample * n) / fh.sample_rate) < 1*u.ns
+
+        # Get reference data.
+        ref_data = fh.read(nsample * n).reshape((-1, n) + fh.sample_shape)
+        # Move fh back for rt to use.
+        fh.seek(0)
+
+        # Check sequential reading.
+        data1 = rt.read()
+        assert rt.tell() == rt.shape[0]
+        assert abs(rt.time - rt.start_time -
+                   rt.shape[0] / rt.sample_rate) < 1*u.ns
+        assert rt.dtype is data1.dtype
+        assert np.all(ref_data == data1)
+
+        # Check seeking and selective decode.
+        rt.seek(-7, 2)
+        assert rt.tell() == rt.shape[0] - 7
+        data2 = rt.read()
+        assert data2.shape[0] == 7
+        assert np.all(data2 == ref_data[-7:])
+        sec_offset = -0.25 * u.ms
+        rt.seek(sec_offset, 'end')
+        assert rt.tell() == rt.shape[0] + self.convert_time_offset(
+            sec_offset, rt.sample_rate)
+        assert rt.tell(unit=u.ms) == (rt.tell() / rt.sample_rate).to(u.ms)
+        current_offset = rt.tell()
+        rt.seek(2, 'current')
+        assert rt.tell() == current_offset + 2
+        time_offset = rt.start_time + 0.13 * u.ms
+        rt.seek(time_offset, 'start')
+        assert rt.tell() == self.convert_time_offset(
+            time_offset - rt.start_time, rt.sample_rate)
+
+        # Check reading to external array.
+        out = np.empty((11,) + rt.sample_shape)
+        rt.seek(0)
+        rt.read(out=out)
+        assert np.all(out == ref_data[:11])
+
+        # Check closing.
+        rt.close()
+        assert fh.closed
```

Comment:
~Probably also should check that `rt.closed` holds.~ Never mind - don't have that in baseband either.